### PR TITLE
Make sure reciever starts before the sender

### PR DESCRIPTION
**Pull request by duglin**
This PR tries to make sure that Receivers are started before Senders since (I think) for some protocols we'll lose messages if the Receiver starts after the Sender sends an Event. And in those cases it'll result in a timeout in the tests as it waits forever.

This will also create two separate clients in `SendReceive` since not all protocols like using the same client for sending and receiving messages concurrently.

### CODE REVIEW

**Comment by duglin** [comment]
I ran the integration tests in a loop for over 12 hours locally w/o any failures, so I'm hopeful this PR helps.
Ping @embano1 @lionelvillard

**Comment by duglin** [comment]
I'll try to rerun the integration tests over and over as time permits today to keep testing, so don't be surprised if it's not completed. I'm ok with merging as long as it doesn't show a failure.

**Comment by embano1** [comment]
> I ran the integration tests in a loop for over 12 hours locally w/o any failures, so I'm hopeful this PR helps.
> Ping @embano1 @lionelvillard

See my feedback. IMHO the `wait` channel stuff can be dropped as it does not really anything. The sleeps might be fixing the issue for you locally. Not sure how much this holds in CI due to the variation/resources in the underlying infra.

**Comment by duglin** [comment]
See https://github.com/cloudevents/sdk-go/actions/runs/6676121613?pr=977
20 runs without a problem. 

The "wait" at least makes sure the first thread starts before the 2nd.

**Comment by embano1** [comment]
> See https://github.com/cloudevents/sdk-go/actions/runs/6676121613?pr=977 20 runs without a problem.

I take that :)

> The "wait" at least makes sure the first thread starts before the 2nd.

Go's preemptive scheduler for goroutines does not provide any guarantee about the order when `receiver.OpenInbound` is executed in the following example. The `time.Sleep` just happens to be enough to give `receiver.OpenInbound` enough time for now, i.e., when there's enough thread capacity, no cpu throttling, etc.

```go
        wait := make(chan bool)

	go func() {
		wait <- true // <- immediately fires in the channel triggering time.Sleep in the other goroutine before receiver.OpenInbound
		err := receiver.OpenInbound(ctx)
		if err != nil {
			messageChan <- receiveMessage{err: err}
		}
		receiver.Close(ctx)
	}()

	// Wait for other thread to start and run OpenInbound + sleep a sec
	// hoping that things will get ready before we call Receive() below
	<-wait
	time.Sleep(time.Second)
```

**Comment by duglin** [comment]
Yes the sleep is for the call to OpenInbound but without the "wait" there's no guarantee that the sleep won't finish before the other thread even starts. That's where the "wait" comes in.

**Comment by embano1** [comment]
> Yes the sleep is for the call to OpenInbound but without the "wait" there's no guarantee that the sleep won't finish before the other thread even starts. That's where the "wait" comes in.

Thread (goroutine) start != it executes completely - Go's scheduler uses user-space scheduling of Goroutines onto 1-to-many OS threads. The scheduler is preemptive, i.e., `time.Sleep` might still execute **before** `err := receiver.OpenInbound(ctx)` depending on the number of OS threads available, concurrent scheduling of goroutines onto threads, CPU throttling (delaying execution), etc.

tl;dr we're lucky it works ;)

**Comment by duglin** [comment]
Let me try this differently... we're saying (almost) the same thing. Yes I agree we're getting lucky because we'll never know when OpenInbound will not only be run but when the stuff in there actually gets things ready. However, there are two major unknowns: 1) the new thread, and 2) OpenInbound. And the "wait" will at least ensure we're past the creation of the new thread. The "sleep" is our wild guess at how long OpenInbound will take. I'm just trying to reduce the unknowns so we don't need to have "sleep" guess for more things than we need to.

27 successful CI runs now!

**Comment by duglin** [comment]
Ok, locally I've been running them for over 24 hours continually w/o any failures, and as part of the PR I've run it 45 times in a row w/o a single failure (see https://github.com/cloudevents/sdk-go/actions/runs/6676121613?pr=977 and see the history drop-down on the right). I think we should merge this puppy and move on!

**Review by embano1 — APPROVED** [review]
👍

File: test/integration/mqtt_paho/mqtt_test.go [hunk 1]
```diff
@@ -37,12 +37,18 @@ func TestSendEvent(t *testing.T) {
...
+		// send events
+		wait := make(chan bool)
+
 		go func() {
 			client, err := cloudevents.NewClient(protocolFactory(ctx, t, topicName))
 			if err != nil {
 				eventChan <- receiveEvent{err: err}
 				return
 			}
+			wait <- true
```
[the remark this answers is not in the document]
Comment (duglin) [reply test/integration/mqtt_paho/mqtt_test.go:51]: Yea, I wasn't as clear as I could have been. See if this PR helps: https://github.com/cloudevents/sdk-go/pull/978
